src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.Fixer.cs

```diff
@@ -0,0 +1,103 @@
+            var diagnostic = context.Diagnostics.FirstOrDefault();
+
+            if (TryParseLocationInfo(diagnostic, DoNotGuardDictionaryRemoveByContainsKey.ConditionalOperation, out var conditionalOperationSpan) &&
+                TryParseLocationInfo(diagnostic, DoNotGuardDictionaryRemoveByContainsKey.ChildStatementOperation, out var childStatementOperationSpan))
+            {
+                context.RegisterCodeFix(
+                    new DoNotGuardDictionaryRemoveByContainsKeyCodeAction(
+                        context.Document,
+                        conditionalOperationSpan,
+                        childStatementOperationSpan),
+                    diagnostic);
+            }
+        }
+
+        private static bool TryParseLocationInfo(Diagnostic diagnostic, string propertyKey, out TextSpan span)
+        {
+            span = default;
+
+            if (!diagnostic.Properties.TryGetValue(propertyKey, out var locationInfo))
+                return false;
+
+            var parts = locationInfo.Split(new[] { DoNotGuardDictionaryRemoveByContainsKey.AdditionalDocumentLocationInfoSeparator }, StringSplitOptions.None);
+            if (parts.Length != 2 ||
+                !int.TryParse(parts[0], out var spanStart) ||
+                !int.TryParse(parts[1], out var spanLength))
+            {
+                return false;
+            }
+
+            span = new TextSpan(spanStart, spanLength);
+            return true;
+        }
+
+        private class DoNotGuardDictionaryRemoveByContainsKeyCodeAction : CodeAction
+        {
+            private readonly Document _document;
+            private readonly TextSpan _conditionalOperationSpan;
+            private readonly TextSpan _childStatementOperationSpan;
+
+            public override string Title { get; }
+
+            public override string EquivalenceKey { get; }
+
+            public DoNotGuardDictionaryRemoveByContainsKeyCodeAction(
+                Document document,
+                TextSpan conditionalOperationSpan,
+                TextSpan childStatementOperationSpan)
+            {
+                _document = document;
+                _conditionalOperationSpan = conditionalOperationSpan;
+                _childStatementOperationSpan = childStatementOperationSpan;
+                EquivalenceKey = DoNotGuardDictionaryRemoveByContainsKey.RuleId;
+                Title = MicrosoftNetCoreAnalyzersResources.DoNotGuardDictionaryRemoveByContainsKeyTitle;
+            }
+
+            protected override async Task<Document> GetChangedDocumentAsync(CancellationToken cancellationToken)
+            {
+                var documentText = await _document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                var childStatementText = documentText.GetSubText(_childStatementOperationSpan);
+                documentText = documentText.Replace(_conditionalOperationSpan, childStatementText.ToString());
+
+                return _document.WithText(documentText);
```

Comment:
Not sure if working with text is the preferred way.

Reply:
Right. I made adjustments to inherit `DocumentChangeAction` and simply replace the nodes instead, which is much simpler:

```csharp
        private static Document ReplaceConditionWithChild(Document document, SyntaxNode root, SyntaxNode conditionalOperationNode, SyntaxNode childOperationNode)
        {
            var newRoot = root.ReplaceNode(conditionalOperationNode, childOperationNode);
            return document.WithSyntaxRoot(newRoot);
        }
```

However, the formatting is off, causing the tests to fail.

```
content of '/0/Test0.cs' did not match. Diff shown with expected as baseline:
 
 using System;
 using System.Collections.Generic;
 namespace Testopolis
 {
     public class MyClass
     {
         private readonly Dictionary<string, string> MyDictionary = new Dictionary<string, string>();
 
         public MyClass()
         {
-            MyDictionary.Remove("Key");
+                MyDictionary.Remove("Key");
         }
     }
 }
```

Is this the expected behavior of a code fix that uses `DocumentChangeAction` (if so, I could adjust the whitespace in my test), or can I do something do auto-format _only_ my change?

Reply:
@chucker Try to call `.WithAdditionalAnnotations(Formatter.Annotation)` on the new node.

Reply:
Yup, that does the trick, thanks!